vulyk/models/task_types.py

```diff
@@ -63,78 +77,201 @@ def export_reports(self, qs=None):
+            # update stats record
+            self._end_work_session(task, user, answer)
+        except ValidationError as err:
+            raise TaskValidationError(err, get_tb())
+        except (OperationError, LookUpError, InvalidQueryError) as err:
+            raise TaskSaveError(err, get_tb())
+
+    def _is_ready_for_autoclose(self, task, answer):
+        """
+        Checks if task could be closed before it
+        Should be overridden if you need more complex logic.
+
+        :param task: an instance of self.task_model model
+        :type task: AbstractTask
+        :param answer: Task solving result
+        :type answer: AbstractAnswer
+
+        :returns: How many identical answers we got
+        :rtype: bool
+        """
+        raise NotImplementedError()
+
+    def _update_task_on_answer(self, task, answer, user):
+        """
+        Sets flag 'closed' to True if task's goal has been reached
+
+        :param task: an instance of self.task_model model
+        :type task: AbstractTask
+        :param answer: Task solving result
+        :type answer: AbstractAnswer
+        :param user: an instance of User model who provided an answer
+        :type user: models.User
+        """
+        task.users_count += 1
+        task.users_processed.append(user)
+
+        if self._is_ready_for_autoclose(answer, task):
+            task.closed = True
+        else:
+            task.closed = task.users_count == self.redundancy
+
+        task.save()
+
+        # get_help that returns help text or template with help. Might be a
+        # property too
+        #
+        # edit, that will return qs or a paginated list of tasks that was already
+        # complete by a given user or None if editing is prohibited for this task
+        # type
+
+    # TODO: make up something prettier than that mess
+    def _start_work_session(self, task, user):
+        """
+        Starts new WorkSession
+
+        :param task: Given task
+        :type task: AbstractTask
+        :param user: an instance of User model who gets new task
+        :type user: models.User
+        """
+        WorkSession.objects.create(user=user, task=task)
+
+    def _end_work_session(self, task, user, answer):
+        """
+        Ends current WorkSession
+
+        :param task: Given task
+        :type task: AbstractTask
+        :param user: an instance of User model who gets new task
+        :type user: models.User
+        :param answer: Given answer
+        :type answer: AbstractAnswer
+
+        :raises: WorkSessionLookUpError - if session was not found
+        """
+        # TODO: store id of active session in cookies or elsewhere
+        rs = WorkSession.objects(user=user, task=task).order_by('-start_time')
+
+        if rs.count() > 0:
+            rs.first() \
+                .update(
+                set__end_time=datetime.now(),
+                set__answer=answer,
+                set__corrections=answer.corrections)
```

Comment:
Whole amount of actions/corrections given by user in this particular answer. could be used for stats and/or empty reports removal.